test/e2e/ingress.go

```diff
@@ -0,0 +1,311 @@
+			msg += fmt.Sprintf("%v\n", f.Name)
+		}
+		return fmt.Errorf("Found forwarding rules:\n%v", msg)
+	}
+	return nil
+}
+
+var _ = Describe("GLBC", func() {
+	// These variables are initialized after framework's beforeEach.
+	var ns string
+	var client *client.Client
+	var responseTimes, creationTimes []time.Duration
+
+	framework := Framework{BaseName: "glbc"}
+
+	BeforeEach(func() {
+		// This test requires a GCE/GKE only cluster-addon
+		SkipUnlessProviderIs("gce", "gke")
+		framework.beforeEach()
+		client = framework.Client
+		ns = framework.Namespace.Name
+		Expect(waitForRCPodsRunning(client, "kube-system", "glbc")).NotTo(HaveOccurred())
+		Expect(checkLeakedResources()).NotTo(HaveOccurred())
+		responseTimes = []time.Duration{}
+		creationTimes = []time.Duration{}
+	})
+
+	AfterEach(func() {
+		framework.afterEach()
+		wait.Poll(lbPollInterval, lbPollTimeout, func() (bool, error) {
+			if err := checkLeakedResources(); err != nil {
+				Logf("Still waiting for glbc to cleanup: %v", err)
+				return false, nil
+			}
+			return true, nil
+		})
+
+		// In most cases slow loadbalancer creation/startup translates directly to
+		// GCE api sluggishness. However this might be because of something the
+		// controller is doing, eg: maxing out QPS by repeated polling.
+		sort.Sort(timeSlice(creationTimes))
+		perc50 := creationTimes[len(creationTimes)/2]
+		if perc50 > avgLBLatency {
+			Failf("Average creation time is too high: %+v", creationTimes)
+		}
+		By(fmt.Sprintf("Loadbalancer creation latencies %+v", creationTimes))
+
+		sort.Sort(timeSlice(responseTimes))
+		perc50 = responseTimes[len(responseTimes)/2]
+		if perc50 > avgLBLatency {
+			Failf("Average startup time is too high: %+v", responseTimes)
+		}
+		By(fmt.Sprintf("Loadbalancer startup latencies %+v", responseTimes))
+	})
+
+	It("should create GCE L7 loadbalancers through the GLBC cluster addon", func() {
+		// Create numApps apps, exposed via numIng Ingress each with 2 paths.
+		// Eg with numApp=10, numIng=5:
+		// apps: {foo-app-(0-10)}
+		// ingress: {foo-app-(0, 2, 4, 6, 8)}
+		// paths:
+		//  ingress foo-app-0:
+		//	  default1.bar.com
+		//	  foo0.bar.com: /foo0
+		//	  foo1.bar.com: /foo1
+		if numApps < numIng {
+			Failf("Need more apps than Ingress")
+		}
+		step := numApps / numIng
+		By(fmt.Sprintf("Creating %d rcs + svc, and %d routes per Ingress", numApps, step))
+		for i := 0; i < numApps; i = i + step {
+			for j := i; j < i+step; j++ {
+				createApp(client, ns, j)
+			}
+			createIngress(client, ns, i, step)
+		}
+
+		ings, err := client.Extensions().Ingress(ns).List(
+			labels.Everything(), fields.Everything())
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, ing := range ings.Items {
+			// Wait for the loadbalancer IP.
+			start := time.Now()
+			address, err := waitForIngressAddress(client, ing.Namespace, ing.Name, lbPollTimeout)
+			Expect(err).NotTo(HaveOccurred())
+			By(fmt.Sprintf("Found address %v for ingress %v, took %v to come online",
```

Comment:
I use By when something of importance has occurred (in this case the loadbalancer acquiring an IP) because it prints out a bold STEP which sort of makes sense, it's the successful completion of a step in the test